Added documentation for the stack bulk actions (#573)
spacelift-io/user-documentation: docs/concepts/stack/stack-bulk-actions.md

 - Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
 - Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.
 - Lock - Allows to lock the selected stacks for exclusive use.
 - Unlock - Allows to unlock the selected locked stacks.
 - Enable - Allows to enable the selected disabled stacks.
 - Disable - Allows to disable the selected stacks, so they will not trigger any runs.
 - Approve - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
 - Reject - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
 - Run task - Allows to [manually trigger a run](../run/task.md) with a custom command.

!!! info
    The UI will only show the actions that can be performed on the selected stacks filtering out the unavailable ones for your convenience.

> ## How to use bulk actions

Author: I'm not sure if that's the right place for this description, as this is going to be common across all bulk actions instances. So would love to get an idea of where I could put it 🤔

Reviewer: I think we've got a few options here:

1. We could have a single page for bulk actions instead of embedding it in the stacks section. It could then have different sections for the different types of actions that can be performed.
2. We could have a bulk actions page that explains the concept in general terms, and then in the stacks page you would just explain the stack-specific actions.
3. Or we just accept the duplication on the grounds that it's simpler for users to have the specific info for the resource type they're viewing in one place. The downside to this for us is more places to maintain if we change bulk actions.

Maybe @jmfontaine has some thoughts about this?

Author: I went with the 2nd option

# Stack bulk actions

Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:

> 

Reviewer: A couple of general thoughts about the action list:

- Maybe a table would look better than a list for this information?
- Should we sort the items in alphabetical order? The reason I say that is the list is relatively long, so it would make it easier to find specific items.

Author: Is there any table that you could point me towards that you pictured? I'm assuming you want two rows, one with the action name and the other with the description? Looking at the other files I've not seen a lot of tables so I just went with the list 🤔.

Author: As for the order, I wanted to keep the similar order as the actions would appear on the bar (from most popular to least), but alphabetical is fine too :)

Reviewer: Sure - we've got one in the self-hosted section for the provider compatibility matrix: https://docs.spacelift.io/self-hosted/latest/vendors/terraform/terraform-provider. I think there might be some elsewhere, but I'm not sure off the top of my head.

Yeah - I was thinking two columns, one for the action name and another for the description, with a row per action.

But please just treat this as a suggestion - if you don't think it's easier to read the information with the table that's fine. I just thought with so many items it might make it easier to scan.

Reviewer: RE the order - that makes perfect sense, and both options are valid. I just know if I'm reading docs like this and searching for something (vs in a UI when I'm actually using it, and it makes sense for the most used options to be sorted first) it's usually easier for me to find if it's alphabetical.

Author: Yeah after the consideration I think I'll pass on the table, for the other bulk actions there's going to be much less items, so not sure if the table will make it clearer. Plus after your suggestions it's a bit more concise and easier to scan through 👌🏻

Reviewer: No problem - sounds good.

> - Confirm - Allows to confirm the blocking run for the selected stacks that are in the `Unconfirmed` state.

Reviewer: ```suggestion
- Confirm - Confirms the blocking run for stacks in the `Unconfirmed` state.
```

> - Discard - Allows to discard the blocking run for the selected stacks that are in the `Unconfirmed` state.

Reviewer: ```suggestion
- Discard - Discards the blocking run for stacks in the `Unconfirmed` state.
```

> - Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.

Reviewer: ```suggestion
- Trigger - [Triggers a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
```

> - Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.

Reviewer: ```suggestion
- Sync commit - Updates the selected stacks to point at the latest HEAD commit for their tracked branches.
```

> - Lock - Allows to lock the selected stacks for exclusive use.

Reviewer: ```suggestion
- Lock - Locks the selected stacks for exclusive use.
```

> - Unlock - Allows to unlock the selected locked stacks.

Reviewer: ```suggestion
- Unlock - Unlocks the selected stacks that are currently locked.
```

> - Enable - Allows to enable the selected disabled stacks.

Reviewer: ```suggestion
- Enable - Enables the selected stacks.
```

> - Disable - Allows to disable the selected stacks, so they will not trigger any runs.

Reviewer: ```suggestion
- Disable - Disables the selected stacks, so they will not trigger any runs.
```

> - Approve - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).

Reviewer: ```suggestion
- Approve - Adds an approval for the selected stacks that require [approval policy](../policy/approval-policy.md) reviews.
```

> - Reject - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).

Reviewer: ```suggestion
- Reject - Adds a rejection for the selected stacks that require [approval policy](../policy/approval-policy.md) reviews.
```

> - Run task - Allows to [manually trigger a run](../run/task.md) with a custom command.

Reviewer: ```suggestion
- Run task - [Manually triggers a run](../run/task.md) with a custom command.
```

!!! info

>     The UI will only show the actions that can be performed on the selected stacks filtering out the unavailable ones for your convenience.

Reviewer: ```suggestion
    The UI will only show the actions that can be performed on the selected stacks, filtering out the unavailable ones for your convenience.
```

## How to use bulk actions

> When you select stacks on the "Stacks" page a floating action bar will appear at the bottom of the screen. This bar will show the number of selected stacks and the available actions.

Reviewer: ```suggestion
After you select one or more stacks, a floating action bar will appear at the bottom of the screen. This bar will show the number of selected stacks and the available actions.
```

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_floating-bar.png)

Actions that are not available for all the selected stacks will be marked with an icon. You can hover over the icon to see how many stacks are going to be affected.
![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_partial-action-floating-bar.png)

If you need a more detailed view of the selected stacks, click the "See details" button, that will open the bulk actions drawer. From the drawer you can also dismiss any stacks you deem unnecessary. Use floating bar for quick actions and use the drawer when you need to be more careful with your selection.
![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_drawer.png)

On the drawer you will see a fine-grained details on which actions are available for each of the selected stacks.
![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_partial-action-drawer.png)

> Once you select action, you will be presented with the confirmation step, that allows you to add additional details (like a note for the Lock action) and confirm the action itself.

Reviewer: ```suggestion
Once you select an action, you will be presented with a confirmation step that allows you to add additional details (like a note for the Lock action) and to confirm the action itself.
```

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_confirm-action-floating-bar.png)

The same view is also available on the drawer, where you can review the applicable and not applicable items again.
![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_confirm-action-drawer.png)

> Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all actions are done otherwise you will stop the execution of the actions on all the selected items. When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).

Reviewer: ```suggestion
Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all of your actions are performed, otherwise any in-progress actions will be stopped.

When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).
```

# Stack bulk actions

> Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:

Reviewer: Maybe move the "When you make the selection you can choose..." sentence into the "Available actions" section. It's just slightly weird to have a `:` and then a heading following it.

- Run task - [Manually triggers a run](../run/task.md) with a custom command.
- Sync commit - Updates the selected stacks to point at the latest HEAD commit for their tracked branches.
- Trigger - [Triggers a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
- Unlock - Unlocks the selected stacks that are currently locked.

!!! info
    The UI will only show the actions that can be performed on the selected stacks, filtering out the unavailable ones for your convenience.

## How to use bulk actions

After you select one or more stacks, a floating action bar will appear at the bottom of the screen. This bar will show the number of selected stacks and the available actions.

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_floating-bar.png)

Actions that are not available for all the selected stacks will be marked with an icon. You can hover over the icon to see how many stacks are going to be affected.

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_partial-action-floating-bar.png)

### Performing actions

> If you need a more detailed view of the selected stacks, click the "See details" button, that will open the bulk actions drawer. From the drawer you can also dismiss any stacks you deem unnecessary. Use floating bar for quick actions and use the drawer when you need to be more careful with your selection.

Reviewer: ```suggestion
If you need a more detailed view of the selected stacks, click the "See details" button, that will open the bulk actions drawer. From the drawer you can also dismiss any stacks you deem unnecessary. Use the floating bar for quick actions and use the drawer when you need to be more careful with your selection.
```

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_drawer.png)

> On the drawer you will see a fine-grained details on which actions are available for each of the selected stacks.

Reviewer: ```suggestion
On the drawer you will see fine-grained details on which actions are available for each of the selected stacks.
```

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_partial-action-drawer.png)

Once you select an action, you will be presented with a confirmation step that allows you to add additional details (like a note for the Lock action) and to confirm the action itself.

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_confirm-action-floating-bar.png)

The same view is also available on the drawer, where you can review the applicable and not applicable items again.

![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_confirm-action-drawer.png)

Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all of your actions are performed, otherwise any in-progress actions will be stopped.

> When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).

Reviewer: ```suggestion
When execution is complete, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).
```

I just felt like the word "action" was repeated too many times here. I think my suggestion is a little clumsy, so if you can think of a better option then feel free.

Author: `when the bulk execution is complete` 🤷🏻 ?

Reviewer: Yeah - that sounds good. :+1:
